### ISSUE TITLE
Optimize ScalarValue list variant using Array trait object

### ISSUE THREAD

**Issue by houqp** [apache/arrow-datafusion#996]
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

This comes out from the arrow2 PR started by @jorgecarleitao. Here is a discussion on why it would be more efficient to model ScalarValue as `dyn Array`: https://github.com/apache/arrow-datafusion/pull/68#discussion_r706565686

**Describe the solution you'd like**

Change ScalarValue's List variant to:

```rust
List(Option<Arc<dyn Array>>, DataType),
```

**Describe alternatives you've considered**

Current way of warping every list value with a ScalarValue enum has a lot of overhead on read/write.

**Additional context**

We should probably wait for the arrow2 PR merge before starting to work on this.

https://github.com/apache/arrow-datafusion/pull/68#discussion_r707044869 has more discussion on how we can handle this in ballista's protobuf ser/de.

**Comment by alamb** [comment]
Closing in favor of   https://github.com/apache/arrow-datafusion/issues/7352 whihc offers a more modern proposal for this same idea. Otherwise known as "its happening!"
